zerver/migrations/0699_stream_subscriber_count_data_migration.py

```diff
@@ -0,0 +1,53 @@
+from django.db import connection, migrations
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from psycopg2 import sql
+
+
+def set_stream_subscribe_count(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "realm")
+    Stream = apps.get_model("zerver", "Stream")
+    Subscription = apps.get_model("zerver", "subscription")
+    Recipient = apps.get_model("zerver", "recipient")
+
+    # Here we compute per-stream subscriber_count in the sub-query which returns a delta_table,
+    # then Update it in-place using that computed value.
+    # The whole query is then executed by one batch per realm.
+    query = sql.SQL(
+        """UPDATE {stream_table} AS stream
+            SET subscriber_count = delta_table.n_subscripers
+            FROM (
+                SELECT recipient.type_id AS stream_id,
+                COUNT(subscription.user_profile_id) AS n_subscripers
+                FROM {subscription_table} AS subscription
+                JOIN {recipient_table} AS recipient
+                    ON subscription.recipient_id = recipient.id
+                WHERE
+                    recipient.type = 2
+                    AND subscription.active = True
+                    AND subscription.is_user_active = True
+                GROUP BY stream_id
+            ) AS delta_table
+            WHERE
+                stream.realm_id = %(realm_id)s
+                AND stream.id = delta_table.stream_id;
+        """
+    ).format(
+        stream_table=sql.Identifier(Stream._meta.db_table),
+        subscription_table=sql.Identifier(Subscription._meta.db_table),
+        recipient_table=sql.Identifier(Recipient._meta.db_table),
+    )
+
+    cursor = connection.cursor()
+    for realm in Realm.objects.all():
+        cursor.execute(query, {"realm_id": realm.id})
```

Comment:
You can't do this all in a single transaction; set `atomic=False` on the Migration below, like 95% of our migrations do.

I recommend reading the last ~20 migrations and improving this one based on what you see there.

Reply:
> You can't do this all in a single transaction
Ok, but why ? It's always good to know the "why" to learn